### PR TITLE
Added a new ResettableInterface and implemented it where possible.

### PR DESCRIPTION
**Pull request by lyrixx**
Added a new ResetableInterface and implemented it where possible.

When one use Monolog in a long process like an AMQP worker with a
`FingersCrossedHandler` or `BufferHandler` there is a drawback: as soon as there
is an AMQP message that generate a log >= error (for example), all next AMQP
messages will output logs, even if theses messages don't generate log where
level >= error.

In the same context there is a drawback for processor that add an UUID to the
logs. The UUID should change for each AMQP messages.

---

This patch address this issue with a new interface: `ResettableInterface` interface.
Side note: `reset()`, `flush()`, `clear()`,  are already used in Monolog. So
basically, one can use the `reset()` on the `Logger` and on some
`Handler`s / `Processor`s.

It's especially useful for

* the `FingersCrossedHandler`: it `close()` the buffer, then it `clear()` the buffer.
* the `BufferHandler`: it `flush()` the buf

### CODE REVIEW

**Comment by lyrixx** [comment]
I forgot to say that we are using this "feature" for years at SensioLabs. But in a very hacky way (with reflection). IMHO, it's better to expose a clean API. That's why I created this PR.

**Comment by theofidry** [comment]
Fortunately in our case such logging is done only when an error occurs, so we can afford to kill the process to circumvent that issue. A clean API to do it would avoid this "hack" or having to rely on reflection 👍

**Comment by lyrixx** [comment]
This PR needs a rebase, but before finishing it I would like to be sure If it will/could be merged.

More over Symfony get a new component: Messenger. So (I hope so) AMQP (or other queue system) will become more standard. So people will expects a nice logging system.

ping @fabpot @nicolas-grekas @sroze 

Next step: wire this restart method after each message processed in the Messenger Component

File: src/Monolog/Processor/UidProcessor.php [hunk 1]
```diff
@@ -43,4 +45,9 @@ public function getUid(): string
     {
         return $this->uid;
     }
+
+    public function restart(int $length = 7)
```
Comment (nicolas-grekas) [inline src/Monolog/Processor/UidProcessor.php]: I'd suggest to store the length as a property instead of adding in here

**Comment by nicolas-grekas** [comment]
> Next step: wire this restart method after each message processed in the Messenger Component

and on container reset

**Comment by stof** [comment]
Why is it in the 2.0 milestone rather than adding this new feature in 1.x (2.0 is kind of stale, and Symfony will have a hard time moving to it due to the non-continuous migration path for handler authors currently)

File: src/Monolog/Logger.php [hunk 2]
```diff
@@ -325,6 +325,21 @@ public function addRecord(int $level, string $message, array $context = []): boo
         return true;
     }
 
+    public function restart($a = null)
```
Comment (stof) [inline src/Monolog/Logger.php]: why having an argument here ?

File: src/Monolog/Processor/UidProcessor.php (hunk 1 above)
Comment (stof) [reply src/Monolog/Processor/UidProcessor.php]: agreed. None of the APIs are calling it with a value anyway, and we want it to be called with the original config.

File: src/Monolog/Processor/UidProcessor.php [hunk 3]
```diff
@@ -43,4 +45,9 @@ public function getUid(): string
...
+    {
+        $this->__construct($length);
```
Comment (stof) [inline src/Monolog/Processor/UidProcessor.php]: I would rather move the uid generation to a private method and call it, rather than reusing `__construct` (which will be necessary anyway when applying the suggestions). I don't want to push people to think that calling `__construct` as a method is a supported usage of the library.

**Comment by lyrixx** [comment]
> Why is it in the 2.0 milestone

I would prefer 1.X yes.

**Comment by stof** [comment]
Well, I think the 2.0 milestone is because this was submitted to master.

**Comment by lyrixx** [comment]
Thanks. I have addressed your comments and now this PR targets the 1.X branch.

File: src/Monolog/RestartableInterface.php [hunk 4]
```diff
@@ -0,0 +1,25 @@
+<?php declare(strict_types=1);
```
Comment (sroze) [inline src/Monolog/RestartableInterface.php]: Is that OK even if Monolog's 1.x's target is PHP 5.3 ?

Comment (lyrixx) [reply src/Monolog/RestartableInterface.php]: oups ;)

**Comment by lyrixx** [comment]
There is a failure on HHVM, but for another test. So this PR is ready.

**Comment by fabpot** [comment]
I would recommend to drop HHVM support as almost all project have done by now (even on 1.x).

**Comment by nicolas-grekas** [comment]
s/RestartableInterface::restart()/ResettableInteraface::reset()/?
then later one add the Symfony interface on the classes here (or as a child of the Symfony interface?)

**Comment by lyrixx** [comment]
Hmm, `reset` is already used in [monolog](https://github.com/Seldaek/monolog/blob/07b6c1c4171a030787f2bc961272b1887032304e/src/Monolog/Handler/FingersCrossedHandler.php#L150) with a different meanings.

IIRC, this is why I choose `restart` instead of `reset`

**Comment by nicolas-grekas** [comment]
a bridge/proxy would do it also

**Comment by Seldaek** [comment]
Hey, this is something I wanted to support in 2.0, as well as having a Logger::close() to be able to close all handlers (as 2.0 has close in the HandlerInterface).

I am not against supporting this in 1.x though, but I wonder if we shouldn't call this `reset` rather than `restart`. The only name conflict is in the FingersCrossedHandler but really the only reason it has a reset() is to do exactly the same thing as this PR intends, so I don't think it's a problem at all, it just generalizes the concept to the rest of Monolog.

**Comment by lyrixx** [comment]
I updated the PR to rename `ResartableInterface` to `ResetableInterface`

**Comment by nicolas-grekas** [comment]
s/ResetableInterface/ResettableInterface

**Comment by lyrixx** [comment]
@nicolas-grekas Thanks ; done

File: src/Monolog/ResettableInterface.php [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+ * file that was distributed with this source code.
+ */
+
+namespace Monolog;
+
+/**
+ * Handler or Processor implementing this interface will be reseted when Logger::reset is called.
+ *
+ * Reseting an Handler or a Processor usually means cleaning all buffers or
+ * reseting in its internal state.
```
Comment (nicolas-grekas) [inline src/Monolog/ResettableInterface.php]: double "t": resetted, resetting

File: tests/Monolog/Handler/FirePHPHandlerTest.php [hunk 6]
```diff
@@ -77,7 +77,7 @@ class TestFirePHPHandler extends FirePHPHandler
 {
     protected $headers = array();
 
-    public static function reset()
+    public static function resetInternalState()
```
Comment (nicolas-grekas) [inline tests/Monolog/Handler/FirePHPHandlerTest.php]: resetStatic?

File: src/Monolog/ResettableInterface.php [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+ * (c) Jordi Boggiano <j.boggiano@seld.be>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Monolog;
+
+/**
+ * Handler or Processor implementing this interface will be reseted when Logger::reset is called.
```
Comment (nicolas-grekas) [inline src/Monolog/ResettableInterface.php]: resetted ;)
Logger::reset()

Comment (lyrixx) [reply src/Monolog/ResettableInterface.php]: I used reset instead of resetted

**Comment by nicolas-grekas** [comment]
What do you think of https://github.com/symfony/symfony/pull/27093#discussion_r201032119?

image

**Comment by Seldaek** [comment]
Thanks!

**Comment by Seldaek** [comment]
Hey guys so just one more thing before I release this in 1.x.. The more I look at it especially in the master branch for 2.0 where I made close() part of the HandlerInterface, it seems to me like close() and reset() do essentially the same thing.. Or at least they should. Do you also see it like that or not? If so I am thinking to just remove close() in 2.0 and make HandlerInterface implement ResettableInterface. I think it'd simplify things, and it resolves what I had in mind for https://github.com/Seldaek/monolog/issues/197 to add Logger::close() to close all handlers as Logger::reset() was part of this PR.

**Comment by nicolas-grekas** [comment]
If we want to identify a difference, I could see this one: reset is meant to put a service back in a reusable state. This does *not* necessarily imply closing connections to remote backends. Actually, when using the same kernel in a loop, having connections already open for next requests should be a significant part of why these apps are faster (ie no `PDO::__construct()` overhead).
Not sure how it applies here, so that's just for thoughts :)

**Comment by Seldaek** [comment]
That makes sense yes, although I don't know how many handlers would actually benefit from this distinction. But then I need to change the implementation a bit, because right now reset() calls close(), so things are even more confusing. 

It seems to me like the main use case for reset() is fingers crossed and buffer handlers to make sure they're cleared in between two requests/jobs in long running apps.

The use case for close() is to do some cleanup at the very end, but also right now that's when we flush buffers and stuff like that, so I guess in the buffer case you'd want to flush on reset instead of close. And then close() would become mostly useless, except sometimes it calls reset() to flush, and in a few rare cases it closes some resource. Makes sense? Mostly thinking out loud here..

**Comment by nicolas-grekas** [comment]
Makes sense to me yes :)

**Comment by sroze** [comment]
Makes sense. Close becomes useless as it’s basically end of the process so
`reset` is enough :)

**Comment by Seldaek** [comment]
OK see https://github.com/Seldaek/monolog/commit/e07c948042045c9acff160648db45e40f75b1c58 for changes/clarifications. Happy to get any feedback on this.

**Comment by Seldaek** [comment]
https://github.com/Seldaek/monolog/commit/b80352368cd79c978a805a76b724618ccb4ba465 is the core of the changes for 1.x but I messed up the commit msg :s
